lib/archive_utils.rb

```diff
@@ -0,0 +1,382 @@
+          else
+            FileUtils.rm_rf out unless File.file? out
+            FileUtils.mkdir_p(File.dirname(out))
+            # skip creating the symlink for now
+          end
+        else
+          puts safe_realdir(out)
+          puts dest
+          puts file
+          puts entry.name
+          raise SafeExtractionError.new(file, entry.name)
+        end
+      end
+      # Now go through again, only for creating the symlinks
+      zf.each do |entry|
+        if entry.symlink?
+          out = File.join(dest, entry.name)
+          link_target = entry.get_input_stream.read
+          # Using realdirpath because symlinks shouldn't need to create any directories
+          if (File.realdirpath(link_target, dest).to_s.starts_with?(dest.to_s) rescue false)
+            File.symlink link_target, out
+          else
+            raise SafeExtractionError.new(file, entry.name)
+          end
+        end
+      end
+    end
+    return true
+  rescue Exception => e
+    raise FileReadError.new(file, 'zip', e)
+  end
+
+  def self.tar_extract(file, dest)
+    File.open(file) do |source| helper_extract(file, source, dest) end
+    return true
+  end
+
+  def self.tar_gz_extract(file, dest)
+    Zlib::GzipReader.open(file) do |source| helper_extract(file, source, dest) end
+    return true
+  end
+
+  def self.gzip_extract(file, dest)
+    Zlib::GzipReader.open(file) do |input_stream|
+      File.open(dest, "w") do |output_stream|
+        IO.copy_stream(input_stream, output_stream)
+      end
+    end
+    return true
+  end
+
+  def self.helper_extract(file, source, destination)
+    # from https://dracoater.blogspot.com/2013/10/extracting-files-from-targz-with-ruby.html
+    Gem::Package::TarReader.new(source) do |tar|
+      dest = nil
+      tar.each do |entry|
+        if entry.full_name == TAR_LONGLINK
+          dest = File.join destination, entry.read.strip
+          next
+        end
+        dest ||= (File.join destination, entry.full_name).sub(/\/$/, "")
+        if (File.realdirpath(dest).to_s.starts_with?(destination.to_s) rescue false)
+          if entry.directory?
+            FileUtils.rm_rf dest unless File.directory? dest
+            FileUtils.mkdir_p dest, :mode => entry.header.mode, :verbose => false
+          elsif entry.file?
+            FileUtils.rm_rf dest unless File.file? dest
+            FileUtils.mkdir_p(File.dirname(dest))
+            File.open dest, "wb" do |f|
+              f.print entry.read
+            end
+            FileUtils.chmod entry.header.mode, dest, :verbose => false
+          elsif entry.header.typeflag == '2' #Symlink!
+            FileUtils.rm_rf dest unless File.file? dest
+            FileUtils.mkdir_p(File.dirname(dest))
+            # skip creating the symlink for now
+          end
+        else
+          raise SafeExtractionError.new(file, entry.full_name)
+        end
+        dest = nil
+      end
+      # Now go through again, only for creating the symlinks
```

Comment:
I'm not sure, honestly.  I'm including them (a) because I can imagine a "basic shell competency" course, where a plausible assignment is to demonstrate that you can construct a particular directory/file structure, and more importantly (b) because I tend to regret not being thorough and having "semantically-complete" implementations later on, when I wonder "why doesn't Thing X work?" and have to reconstruct the engineering reasons behind the implementation's limits.

I can clean up the efficiency of it, a bit, by setting a flag saying "yes, we've seen a symlink", and only traversing the archive again if there were any symlinks to be handled.